File: word_maker WatchKit Extension/WordHelper.swift
//
//  WordHelper.swift
//  word_maker WatchKit Extension
//
//  Created by MacBook Air on 2022-11-26.
//

import Foundation

public class WordHelper {    
    private static var wordHelper = WordHelper()
    
    public static func instance() -> WordHelper {
        return wordHelper
    }
    
    var words: [[String]] = [[], [], [], [], [], [],[], [], []]
    
    public func getWords() -> [String] {
        return self.words[self.bookIndex];
    }
    
    private init() {
        for index in 3-3 ... 9-3 {
            let lines = self.loadWords(index: index)
            self.words[index] = lines
        }
        self.numbers = self.loadNumbers()
    }
    
    private var bookIndex = -1
    
    public func getBookIndex() -> Int {
        return self.bookIndex
    }
    
    public func setBookIndex(_ bookIndex: Int) {
        self.bookIndex = bookIndex
        self.wordIndex = self.numbers[bookIndex]
    }
    
    private var wordIndex = 0
    
    public func getWordIndex() -> Int {
        return self.wordIndex
    }
    
    public func getWordCount() -> Int {
        return self.words[self.bookIndex].count
    }
    
    public func setWordIndex(next: Bool) {
        if(next) {
            self.wordIndex = self.wordIndex + 1
            if(self.wordIndex >= self.words[self.bookIndex].count) {
                self.wordIndex = self.words[self.bookIndex].count - 1
            }
            self.numbers[self.bookIndex] = self.wordIndex
            self.saveNumbers(numbers: self.numbers)
        }
    }
    
    public func getWord() -> String {
        self.words[self.bookIndex][self.wordIndex]
    }
    
    public func loadWords(index: Int) -> [String] {
        var lines: [String] = []
        do {
            let name: String = "word_" + String(index + 3) + ".txt"
            let path = Bundle.main.path(forResource: name, ofType: nil)
            let file = try String(contentsOfFile: path!)
            lines = file.components(separatedBy: "\n")
        } catch let error {
            Swift.print("Fatal Error: \(error.localizedDescription)")
        }
        return lines
    }
    
    var numbers: [Int] = [0, 0, 0, 0, 0, 0, 0, 0, 0]
    
    public func loadNumbers() -> [Int] {
        var numbers: [Int] = [0, 0, 0, 0, 0, 0, 0, 0, 0]
        do {
            let name: String = "word_number.txt"
            let paths = FileManager.default.urls (for: .documentDirectory, in: .userDomainMask)
            let path = paths[0].appendingPathComponent(name)
            let file = try String(contentsOf: path)
            let lines = file.components(separatedBy: "\n")
            var index = 0
            for line in lines {
                numbers[index] = Int(line) ?? 0
                index = index + 1
            }
        } catch let error {
            Swift.print("Fatal Error: \(error.localizedDescription)")
        }
        return numbers
    }
    
    public func saveNumbers(numbers: [Int]) {
        do {
            let name: String = "word_number.txt"
            let paths = FileManager.default.urls (for: .documentDirectory, in: .userDomainMask)
            let path = paths[0].appendingPathComponent(name)
            var strings = [String]()
            for number in numbers {
                strings.append(String(number))
            }
            let text = strings.joined(separator: "\n")
            try text.write(to: path, atomically: false, encoding: .utf8)
        } catch let error {
            Swift.print("Fatal Error: \(error.localizedDescription)")
        }
    }
}

File: app/src/main/java/com/jchip/word/maker/viewer/WordHelper.java
package com.jchip.word.maker.viewer;

import android.content.Context;
import android.util.Log;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.List;

public class WordHelper {
    private static WordHelper wordHelper = new WordHelper();

    public static WordHelper instance() {
        return wordHelper;
    }

    List<List<String>> words;
    List<Integer> numbers;

    private int bookIndex = -1;
    private int wordIndex = 0;

    public List<String> getWords() {
        return this.words.get(this.bookIndex);
    }

    private WordHelper() {
    }


    public int getBookIndex() {
        return this.bookIndex;
    }

    public void setBookIndex(int bookIndex) {
        this.bookIndex = bookIndex;
        this.wordIndex = this.numbers.get(bookIndex);
    }


    public int getWordIndex() {
        return this.wordIndex;
    }

    public int getWordCount() {
        return this.words.get(this.bookIndex).size();
    }

    public void setNextWord() {
        this.wordIndex = this.wordIndex + 1;
        if (this.wordIndex >= this.words.get(this.bookIndex).size()) {
            this.wordIndex = this.words.get(this.bookIndex).size() - 1;
        }
        this.numbers.set(this.bookIndex, this.wordIndex);
    }

    public String getWord() {
        return this.words.get(this.bookIndex).get(this.wordIndex);
    }

    public void loadWords(Context context) {
        if (this.words == null) {
            this.words = new ArrayList<>();
            for (int index = 3; index <= 9; index++) {
                this.words.add(this.loadWords(context, index));
            }
        }
        if (this.numbers == null) {
            this.loadNumbers(context);
        }
    }

    public List<String> loadWords(Context context, int index) {
        List<String> lines = new ArrayList<>();
        try {
            int id = context.getResources().getIdentifier("word_" + index, "raw", context.getPackageName());
            InputStream file = context.getResources().openRawResource(id);
            BufferedReader reader = new BufferedReader(new InputStreamReader(file));
            while (true) {
                String line = reader.readLine();
                if (line == null) {
                    break;
                }
                lines.add(line);
            }
        } catch (Exception error) {
            Log.d("Fatal Error: ", error.toString());
        }
        return lines;
    }

    public void loadNumbers(Context context) {
        this.numbers = new ArrayList<>();
        try {
            String name = "word_number.txt";
            String path = context.getFilesDir().getAbsolutePath() + File.separator + name;
            InputStream file = new FileInputStream(path);
            BufferedReader reader = new BufferedReader(new InputStreamReader(file));
            String line;
            do {
                line = reader.readLine();
                this.numbers.add(Integer.parseInt(line));
            } while (line != null);
            reader.close();
        } catch (Exception error) {
            Log.d("Fatal Error: ", error.toString());
        }
        for (int index = this.numbers.size(); index < 9; index++) {
            this.numbers.add(0);
        }
    }

    public void saveNumbers(Context context) {
        try {
            String name = "word_number.txt";
            String path = context.getFilesDir().getAbsolutePath() + File.separator + name;
            OutputStream file = new FileOutputStream(path);
            BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(file));
            for (int number : this.numbers) {
                writer.write(String.valueOf(number));
                writer.newLine();
            }
            writer.close();
        } catch (Exception error) {
            Log.d("Fatal Error: ", error.toString());
        }
    }
}
